source/app.d

```diff
@@ -8,33 +8,150 @@ import html;
 
 import output;
 
-void main() {
-	Func[] funcs;
-
-	auto output = File("testoutput.d", "w");
-	auto ltw = output.lockingTextWriter();
-
-	auto doc = createDocument(readText("data-3.4.4.xml"));
-	//writeln(doc);
-	foreach(it; doc.querySelectorAll("intrinsic")) {
-		Func cur;
-		cur.name = to!string(it.attr("name"));
-		cur.ret = Param("", to!string(it.attr("rettype")));
-		cur.tech = to!string(it.attr("tech"));
-		auto t = doc.querySelector("operation", it);
-		if(t is null) {
-			continue;
-		}
-		cur.operation = to!string(t.html());
-		foreach(p; doc.querySelectorAll("parameter", it)) {
-			cur.params ~= Param(
-					to!string(p.attr("varname")),
-					to!string(p.attr("type"))
-				);
-		}
-		toD(ltw, cur);
-		//writeln("\n", toD(cur));
-		funcs ~= cur;
-	}
-	//writefln("%(%s\n%)", funcs);
+void usage()
+{
+    writeln;
+    writeln("Usage: intrinsicripper [-i data-X.Y.Z.xml] [-o output.d] [-t MMX] [-h] [-b]");
+    writeln();
+    writeln("Params:");
+    writeln("  -i          Specify an input intrinsics XML file (default: \"data-3.4.4.xml\")");
+    writeln("  -o          Specify an output D file (default: stdout)");
+    writeln("  -n, --name  Specify a particular intrinsic name (default: unused)");
+    writeln("  -t, --tech  Specify a tech selector (default: none)");
+    writeln("              Available techs: ALL MMX SSE SSE2 SSE3 SSSE3 SSE4.1 SSE4.2...");
+    writeln("  -h, --help  Display this help");
+    writeln("  -b, --body  generate D bodies converted from Intel pseudo-code (default: no).");
+    writeln;
+}
+
+int main(string[] args) 
+{
+    try
+    {
+        string inputXML = "data-3.4.4.xml";
+        string outputPath = null;
+
+        // selected sets is the union of selected instruction sets and selected names
+        string[] selectedTech = [];
+        string[] selectedNames = [];
+        bool help = false;
+        bool generateBodies = false;
+
+        for(int i = 1; i < args.length; ++i)
```

Comment:
please have a look at std.getopt

Reply:
well I don't see how is that a problem

Reply:
not a problem but you can get rid of half of the lines of code and can replace this with a very well tested piece of library.